colexec: HashRouter fixes

asubiotto opened the pull request:
This commit renames routerOutputOp.reset to resetForBenchmarks to make it
explicit that this method is to be called only during benchmarks. The method
could previously be called during normal execution since it implemented the
`resetter` interface.

Unlocks in the routerOutputOp are now called in defers, to ensure that if a
panic occurs (which is normal during vectorized execution), the lock is
unlocked properly.

Release note (bug fix): possible deadlock during vectorized query execution has
been fixed

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49106)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49106#-:-M7NoL8WB_Q60ah7UBhq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
colexec: HashRouter fixes

This commit renames routerOutputOp.reset to resetForBenchmarks to make it
explicit that this method is to be called only during benchmarks. The method
could previously be called during normal execution since it implemented the
`resetter` interface.

Unlocks in the routerOutputOp are now called in defers, to ensure that if a
panic occurs (which is normal during vectorized execution), the lock is
unlocked properly.

Release note (bug fix): possible deadlock during vectorized query execution has
been fixed

asubiotto commented:
bors r=yuzefovich
